src/Fmt_ast.ml

```diff
@@ -303,18 +303,41 @@ let rec fmt_longident (li: Longident.t) =
       cbox 0 (fmt_longident li1 $ wrap "(" ")" (fmt_longident li2))
 
 
-let fmt_constant const =
+let fmt_constant (conf: Conf.t) const =
+  let ocaml_whitespace = function
+    | '\n' | '\t' | '\r' | ' ' -> true
+    | _ -> false
+  in
+  let escape_char c =
+    let code = Char.to_int c in
+    match conf.Conf.escape_chars with
+    | _ when ocaml_whitespace c || Char.equal c '\\' -> str (Char.escaped c)
+    | _ when 0x20 <= code && code <= 0x7E -> char c
+    | `Hexadecimal -> str (Printf.sprintf "\\x%02x" code)
+    | `Minimal -> char c
+    | `Octal -> str (Char.escaped c)
+  in
   match const with
   | Pconst_integer (lit, suf) | Pconst_float (lit, suf) ->
       str lit $ opt suf char
-  | Pconst_char c -> wrap "'" "'" @@ str (Char.escaped c)
+  | Pconst_char c -> wrap "'" "'" @@ escape_char c
   | Pconst_string (s, delim) ->
+      let escape_literal string =
+        String.fold string ~init:(false, fmt "@[") ~f:
+          (fun (freshline, prev) ch ->
+            match (ch, conf.Conf.break_string_literals) with
+            | ' ', _ when freshline -> (false, prev $ str "\\ ")
+            | '\n', `New_lines -> (true, prev $ fmt "\\n\\@\n")
```

Comment:
I think that this case will apply for e.g. format strings such at the one on this line, which is probably not desirable since in a format string `@\n` is semantically an atomic unit.

Reply:
OK, not sure that is what you mean, but I'll try not outputing `@\n` when the `\n` is the last character

To get

```
  let x fmt = Format.fprintf fmt "@[Hello@\n\
                                  World@]@\n"
```

instead of

```
  let x fmt = Format.fprintf fmt "@[Hello@\n\
                                  World@]@\n\
                                  "
```